Mathlib/Algebra/Polynomial/SumIteratedDerivative.lean

```diff
@@ -0,0 +1,260 @@
+/-
+Copyright (c) 2022 Yuyang Zhao. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuyang Zhao
+-/
+import Mathlib.Algebra.Polynomial.AlgebraMap
+import Mathlib.Algebra.Polynomial.BigOperators
+import Mathlib.Algebra.Polynomial.Degree.Lemmas
+import Mathlib.Algebra.Polynomial.Derivative
+import Mathlib.Algebra.Polynomial.RingDivision
+
+/-!
+# Sum of iterated derivatives
+-/
+
+open Finset
+open scoped Nat
+
+namespace Polynomial
+
+variable {R : Type*}
+
+section Semiring
+
+variable [Semiring R]
+
+theorem sum_iterate_derivative_apply_of_lt {p : R[X]} {n : ℕ} (hn : p.natDegree < n) :
+    ∑ i ∈ range (p.natDegree + 1), derivative^[i] p = ∑ i ∈ range n, derivative^[i] p := by
+  obtain ⟨m, rfl⟩ := Nat.exists_eq_add_of_lt hn
+  rw [add_right_comm, sum_range_add _ _ m]
+  convert (add_zero _).symm
+  apply sum_eq_zero
+  intro x _
+  rw [add_comm, Function.iterate_add_apply, iterate_derivative_eq_zero (lt_add_one _),
+    iterate_derivative_zero]
+
+theorem sum_iterate_derivative_apply_of_le {p : R[X]} {n : ℕ} (hn : p.natDegree ≤ n) :
+    ∑ i ∈ range (p.natDegree + 1), derivative^[i] p = ∑ i ∈ range (n + 1), derivative^[i] p :=
+  sum_iterate_derivative_apply_of_lt (Nat.lt_add_one_iff.mpr hn)
+
+/--
+Sum of iterated derivatives of a polynomial, as a linear map
+
+This definition does not allow different weights for the derivatives. It is likely that it could be
+extended to allow them, but this was not needed for the initial use case (the integration by part
+of the integral $I_i$ in the
+[Lindemann-Weierstrass](https://en.wikipedia.org/wiki/Lindemann%E2%80%93Weierstrass_theorem)
+theorem).
+-/
+noncomputable def sumIderiv : R[X] →ₗ[R] R[X] where
+  toFun p := ∑ i ∈ range (p.natDegree + 1), derivative^[i] p
```

Comment:
Thinking again;what do you think of first defining
```
def derivativeFinsupp : R[X] →ₗ[R] ℕ →₀ R[X]
```
with `derivFinsupp p i = derivative^[i] p`? this function then falls out by composing with `Finsupp.lsum`.